libgnucash/engine/test/test-guid.cpp

```diff
@@ -65,24 +67,26 @@ run_test (void)
     book = qof_session_get_book (sess);
     do_test ((NULL != book), "book not created");
 
+    std::vector<std::unique_ptr<void,void(*)(void*)>> ents;
```

Comment:
Well, cute is one way of saying it. I thought it was rather ugly.
Good to know that writing C code in .cpp files is not frowned upon. I'll give your suggestion a try.

And as far as I can tell, the collection never takes ownership of the objects that are added, they remain the responsibility of the caller. In the case of this test they do need to be cleaned up in the test, but for normal application code I would presume the objects are owned by the book or session in the first place.

Reply:
Most of the *.cpp files started out as C files and got converted to use some C++ thing in a function. If they'd started as C++ we sure wouldn't be using GLib containers!

> the collection never takes ownership

That's correct and is exactly the problem I'm referring to. The collections are loaded at session startup from the backend and any QofInstances created with the UI are added; only instances explicitly deleted in the session are freed (maybe). When the session ends the book is destroyed and it calls `qof_collection_destroy` without freeing the QofInstances explicitly. If that weren't true then the `qof_session_destroy` would have cleaned everything up.

The other part of the problem is that a lot of the QofInstance classes aren't correctly implemented as GObject subclasses so calling `g_object_unref` on them might not get all of their bits--or in some cases any of their bits--freed. That turns into every collection would need a type-specific destruction function passed to `g_hash_table_new_full` so that it could clean itself up correctly.

We decided when we embarked on the C++-ification that we'd defer fixing that until we were in a position to rewrite the whole thing as C++ objects with correct memory management. I don't think anyone thought that it would take 10 years and counting to get there.